SSA/Projects/FullyHomomorphicEncryption/Basic.lean

```diff
@@ -81,72 +142,240 @@ abbrev R := (ZMod q)[X] ⧸ (Ideal.span {f q n})
 -- Coefficients of `a : R' q n` are `a\_i : Zmod q`.
 -- TODO: get this from mathlib
 
-/-- Canonical epimorphism `ZMod q[X] ->*+ R q n` -/
+/-- Canonical epimorphism / quotient map: `ZMod q[X] ->*+ R q n` -/
 abbrev R.fromPoly {q n : Nat} : (ZMod q)[X] →+* R q n := Ideal.Quotient.mk (Ideal.span {f q n})
 
-noncomputable abbrev R.zero {q n : Nat} : R q n := R.fromPoly 0
-noncomputable abbrev R.one {q n : Nat} : R q n := R.fromPoly 1
+/-- fromPoly, the canonical epi from `ZMod q[X] →*+ R q n` is surjective -/
+theorem R.surjective_fromPoly (q n : ℕ) : Function.Surjective (R.fromPoly (q := q) (n := n)) := by
+  simp[R.fromPoly]
+  apply Ideal.Quotient.mk_surjective
+
+private noncomputable def R.representative' : R q n → (ZMod q)[X] := Function.surjInv (R.surjective_fromPoly q n)
+
+theorem R.injective_representative' (q n : ℕ) : Function.Injective (R.representative' (q := q) (n := n)) := by
+  simp[R.representative']
+  apply Function.injective_surjInv
+
+/-- A concrete version that shows that mapping into the ideal back from the representative produces the representative'
+  NOTE: Lean times out if I use the abbreviation `R.fromPoly` for unclear reasons! -/
+theorem R.fromPoly_representatitive' (a : R q n) : R.fromPoly (R.representative' q n a) = ↑ a := by
+  simp[R.fromPoly, R.representative']
+  apply Function.surjInv_eq
 
-noncomputable instance {q n : Nat} : Zero (R q n) := ⟨R.zero⟩
-noncomputable instance {q n : Nat} : One (R q n) := ⟨R.one⟩
+theorem R.fromPoly_representatitive'_toFun (a : R q n) : (R.fromPoly (q := q) (n := n)).toFun (R.representative' q n a) = ↑a := by
+  apply Function.surjInv_eq
 
-private noncomputable def R.representative' : R q n → (ZMod q)[X] := Function.surjInv Ideal.Quotient.mk_surjective
 /--
 The representative of `a : R q n` is the (unique) polynomial `p : ZMod q[X]` of degree `< 2^n`
  such that `R.fromPoly p = a`.
 -/
 noncomputable def R.representative : R q n → (ZMod q)[X] := fun x => R.representative' q n x %ₘ (f q n)
 
+@[simp]
+theorem R.fromPoly_kernel_eq_zero (x : (ZMod q)[X]) : R.fromPoly (n := n) (f q n * x) = 0 := by
```

Comment:
Yould just prove that `R.fromPoly (f q n) = 0` and then simp would prove this from ringhom properties and ring properties.